wsd/ClientRequestDispatcher.cpp

```diff
@@ -888,6 +888,62 @@ void ClientRequestDispatcher::handleIncomingMessage(SocketDisposition& dispositi
 #endif // MOBILEAPP
 }
 
+namespace
+{
+
+bool allowedOriginByHost(const std::string& host, const std::string& actualOrigin)
+{
+    // always allow https host to match origin
+    if (net::sameOrigin("https://" + host, actualOrigin))
+        return true;
+    // allow http too if not ssl enabled
+    if (!ConfigUtil::isSslEnabled() && net::sameOrigin("http://" + host, actualOrigin))
+        return true;
+    return false;
+}
+
+bool allowedOrigin(const Poco::Net::HTTPRequest& request, const RequestDetails& requestDetails)
+{
+    const std::string actualOrigin = request.get("Origin");
+
+    const ServerURL cnxDetails(requestDetails);
+
+    if (net::sameOrigin(cnxDetails.getWebServerUrl(), actualOrigin))
+    {
+        LOG_TRC("Allowed Origin: " << actualOrigin << " to match " << cnxDetails.getWebServerUrl());
+        return true;
+    }
+
+#if !MOBILEAPP
+    if (COOLWSD::IndirectionServerEnabled && COOLWSD::GeolocationSetup)
+    {
+        if (HostUtil::allowedWSOrigin(actualOrigin))
+        {
+            LOG_TRC("Allowed Origin: " << actualOrigin << " to match AllowedWSOriginList");
+            return true;
+        }
+    }
+#endif
+
+    const std::string host = request.get("Host");
+    if (allowedOriginByHost(host, actualOrigin))
+    {
+        LOG_DBG("Allowed Origin: " << actualOrigin << " to match against host: " << host);
+        return true;
+    }
+    if (host != COOLWSD::ServerName && !COOLWSD::ServerName.empty() &&
```

Comment:
So this will allow for cases where normally we are sitting behind a terminating proxy, but we are invoked directly (as we suspect is happening with the admin console) ? Makes sense.

Reply:
yes, in this case it was a "Rejecting WebSocket upgrade with: origin [http://something:9980/] expected [https://something:9980/] instead" so I believe the admin console was accessed directly with http:// with ssl termination was enabled in coolwsd.xml